pkg/util/tracing.go

```diff
@@ -0,0 +1,96 @@
+package util
+
+import (
+	"sync"
+	"time"
+
+	"go.opencensus.io/trace"
+)
+
+// ConstantRateTraceSampler uses a token bucket algorithm to sample traces. This is useful as an alternative
+// to probability-based sampling when the service has a low number of requests per second.
+//
+// Inspired in part by https://github.com/jaegertracing/jaeger-client-go/blob/master/utils/rate_limiter.go
+
+type ConstantRateTraceSampler struct {
+	lock sync.Mutex
+
+	// The number of available tokens in the bucket.
+	AvailableTokens int64
+
+	// Maximum number of tokens in the bucket.
+	MaxTokens int64
+
+	// Size of a period in nanoseconds.
+	PeriodDurationNanos int64
+
+	// Number of tokens put into the bucket per time period.
+	TokensPerPeriod int64
+
+	// Number of tokens required to sample a trace.
+	TokensPerSample int64
+
+	LastRefillTime time.Time
+	NextRefillTime time.Time
+	ClockNow       func() time.Time
```

Comment:
Be sure to use this interface instead:
https://pkg.go.dev/github.com/buildbarn/bb-storage@v0.0.0-20210207101039-9507e33a5caf/pkg/clock#Clock
We already create mocks for it in internal/mock, so inside your tests you can just say
```go
clock := mock.NewMockClock(ctrl)
clock.EXPECT().Now().Return(time.Unix(1000, 0))
```

Reply:
Will do. That might obviate the need for some of the field visibility issues.